File: ios_app/iphone/TaskFormSheet.swift
import SwiftUI
import shared

struct TaskFormSheet: View {

    @State private var vm: TaskFormSheetVm
    @Binding private var isPresented: Bool

    init(
            task: TaskDb?,
            isPresented: Binding<Bool>
    ) {
        _vm = State(initialValue: TaskFormSheetVm(task: task))
        _isPresented = isPresented
    }

    var body: some View {

        VMView(vm: vm, stack: .VStack(spacing: 0)) { state in

            SheetHeaderView(
                    onCancel: { isPresented.toggle() },
                    title: state.headerTitle,
                    doneText: state.headerDoneText,
                    isDoneEnabled: state.isHeaderDoneEnabled,
                    scrollToHeader: 0
            ) {
                vm.save {
                    isPresented = false
                }
            }

            MyListView__ItemView(
                    isFirst: true,
                    isLast: true
            ) {

                MyListView__ItemView__TextInputView(
                        text: state.inputTextValue,
                        placeholder: "Task",
                        isAutofocus: true
                ) { newValue in
                    vm.setInputTextValue(text: newValue)
                }
            }
                    .padding(.top, 10)

            MyListView__Padding__SectionSection()

            TextFeaturesTimerFormView(
                    textFeatures: state.textFeatures
            ) { textFeatures in
                vm.setTextFeatures(newTextFeatures: textFeatures)
            }

            MyListView__Padding__SectionSection()

            TextFeaturesTriggersFormView(
                    textFeatures: state.textFeatures
            ) { textFeatures in
                vm.setTextFeatures(newTextFeatures: textFeatures)
            }

            Spacer()
        }
                .background(c.sheetBg)
    }
}

File: android_app/src/main/java/me/timeto/app/ui/TaskFormSheet.kt
package me.timeto.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalSoftwareKeyboardController
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.unit.dp
import me.timeto.app.c
import me.timeto.app.rememberVm
import me.timeto.shared.db.TaskDb
import me.timeto.shared.vm.TaskFormSheetVm

@Composable
fun TaskFormSheet(
    task: TaskDb?,
    layer: WrapperView.Layer,
) {

    val (vm, state) = rememberVm(task) { TaskFormSheetVm(task) }

    val keyboardController = LocalSoftwareKeyboardController.current

    Column(
        modifier = Modifier
            .background(c.sheetBg)
            .navigationBarsPadding()
            .imePadding(),
    ) {

        val scrollState = rememberScrollState()

        Sheet.HeaderViewOld(
            onCancel = { layer.close() },
            title = state.headerTitle,
            doneText = state.headerDoneText,
            isDoneEnabled = state.isHeaderDoneEnabled,
            scrollState = scrollState,
        ) {
            vm.save {
                layer.close()
            }
        }

        Column(
            modifier = Modifier
                .verticalScroll(state = scrollState)
                .padding(bottom = 20.dp)
                .navigationBarsPadding()
                .imePadding()
        ) {

            MyListView__ItemView(
                isFirst = true,
                isLast = true,
            ) {

                MyListView__ItemView__TextInputView(
                    placeholder = "Task",
                    text = state.inputTextValue,
                    onTextChanged = { vm.setInputTextValue(it) },
                    isAutofocus = true,
                    keyboardButton = ImeAction.Done,
                    keyboardEvent = { keyboardController?.hide() },
                )
            }

            MyListView__Padding__SectionSection()

            TextFeaturesTimerFormView(state.textFeatures) {
                vm.setTextFeatures(it)
            }

            MyListView__Padding__SectionSection()

            TextFeaturesTriggersFormView(state.textFeatures) {
                vm.setTextFeatures(it)
            }
        }
    }
}
